### PR TITLE
Img Component

### PR DESCRIPTION
**Pull request by JamieB-gu**
## What does this change?

First PR to this repo, creating the first component, `Img`, designed to render `` tags. Also introduces the `package.json`, `package-lock.json` and `tsconfig.json` files required to manage a TypeScript project.

The short-term goal of this component is for apps-rendering to be able to require it as a dependency, used to render images. Longer-term we would like to have dotcom-rendering depend on this too.

I'll follow up by adding tests and storybook, but haven't done it here in the interests of keeping this PR shorter.

- `Img` component to render `` tags
- Created `package.json`
- Added dependencies required for component
- Added `tsconfig`
- Added `image.ts` module to manage the `Image` type
- Added `lib.ts` module to hold the tagged template function for dark mode

FYI @SiAdcock 

## How can we measure success?

This component can be used to render images in apps-rendering, and eventually in dotcom-rendering.

### CODE REVIEW

**Comment by gtrufitt** [comment]
👀 Niiice. CC/ @mxdvl who had an interest

**Comment by paperboyo** [comment]
cc @philmcmahon

File: src/components/img.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        />
+        <img
+            src={image.src}
+            alt={withDefault('')(image.alt)}
+            className={getLightboxClassName(image.width, lightboxClassName)}
+            css={[
+                styles(image.role, format, supportsDarkMode),
+                withDefault<SerializedStyles | undefined>(undefined)(className),
+            ]}
+            data-ratio={image.height / image.width}
```
Comment (mxdvl) [inline src/components/img.tsx:85]: Might be worth implementing some `padding-top: ${100/ratio}%` ?

https://css-tricks.com/aspect-ratio-boxes/

Comment (JamieB-gu) [reply src/components/img.tsx:85]: Interesting. Is there a reason that's preferable to doing:

```ts
height: ${100/ratio}%;
```

or even:

```ts
height: calc(100% * ${ratio});
```

?

Also worth noting that this component may typically be used by a more specific image component that's more prescriptive about the styles, e.g. [BodyImage](https://github.com/guardian/apps-rendering/blob/d1e6268b9ddcc28e0458be6c248ca7fda404f474/src/components/bodyImage.tsx#L25), which I plan to implement next. That said, it may be preferable to always set the `height` on the `Img` component using percentages, because it's unlikely that we'll want the ratio to change.

Comment (JamieB-gu) [reply src/components/img.tsx:86]: I've pushed a change to specify height for now.

Comment (mxdvl) [reply src/components/img.tsx:86]: Yes, setting width and height is ideal, but if for some reason the images end up being rendered in the viewport with different dimensions, that's where a ratio-box is useful.

In this case, it would actually imply wrapping the whole `picture` element in a relatively positioned `div`, which would have this padding. It's far from clean, but until the [aspect-ratio css property is properly supported][ar], that's [the best we have][sm] on the web. 

[sm]: https://www.smashingmagazine.com/2020/03/setting-height-width-images-important-again/
[ar]: https://caniuse.com/#feat=mdn-css_properties_aspect-ratio

File: src/components/img.tsx [hunk 2]
```diff
@@ -47,18 +47,19 @@ interface Props {
...
-    switch (role) {
+    switch (image.role) {
         case Role.Thumbnail:
             return css`color: ${neutral[60]};`;
         default:
             return css`
                 background-color: ${backgroundColour(format)};
                 color: ${neutral[60]};
                 ${darkModeCss(supportsDarkMode)`background-color: ${neutral[20]};`}
+                height: ${100 * image.height / image.width}%;
```
Comment (mxdvl) [inline src/components/img.tsx:62]: This won't work. [Padding][] percentages are always based on the **`width`** of the containing block, but `height` is not

[Padding]: https://developer.mozilla.org/en-US/docs/Web/CSS/padding#Values

File: src/components/img.tsx (hunk 1 above)
Comment (JamieB-gu) [reply src/components/img.tsx:86]: This component should ideally be context-agnostic, there's no assumption that it *will* live inside a `picture` element, so in this case you'd have to wrap it directly in a `div`.

File: src/components/img.tsx (hunk 2 above)
Comment (JamieB-gu) [reply src/components/img.tsx:62]: Good point, that's what you get for developing without actually getting a browser involved 🙈.

I'm tempted to wait to implement this until at least Storybook is up-and-running, and perhaps one or two of the consumer components as well 🤔. It may be that we'll always be able to specify the height from context, as we do [here](https://github.com/guardian/apps-rendering/blob/d1e6268b9ddcc28e0458be6c248ca7fda404f474/src/components/bodyImage.tsx#L25) for body images.

Comment (mxdvl) [reply src/components/img.tsx]: These body images are also using a mix of percentage (`%`) and viewport-width (`vw`) values, which is probably not ideal.

Comment (JamieB-gu) [reply src/components/img.tsx]: > which is probably not ideal

Can you elaborate on why this might be an issue?

Comment (mxdvl) [reply src/components/img.tsx]: The percentage value is based on the [containing block][], whereas the viewport is based on the device. This means that 100% will only ever be the same as 100% if the containing block is taking up the full width of the viewport.

[containing block]: https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_Block

Comment (JamieB-gu) [reply src/components/img.tsx]: Interesting point. I think previously the width _was_ `100%` of the viewport because on the app images bleed to the edge of the screen. But since dotcom doesn't do that, and we're trying to avoid design discrepancies with dotcom, it was changed to have the same padding as the text.

This _has_ caused an aspect ratio issue as you mentioned, but I suspect it wasn't noticed because of the `object-fit: cover` which [crops anything](https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit) that doesn't fit into the content box.

In summary, you're right and we should change this.

**Comment by paperboyo** [comment]
Please excuse the layman, but is this ever going to be used in the scenario where not only width but also height is fluid (eg restricted by current viewport), like eg in the Lightbox? If yes, then it may be related to [this issue](https://github.com/guardian/frontend/issues/12216). Specifically, [this comment](https://github.com/guardian/frontend/issues/12216#issuecomment-196869603).

**Comment by AWare** [comment]
The `Image` type, is that supposed to eventually come from the CAPI typedefs?

**Comment by JamieB-gu** [comment]
> eg in the Lightbox

Short answer is I don't know. AFAIK each platform does Lightbox quite differently (e.g. web modal on dotcom, native screen on apps); unless we decide to change that and (e.g. switch to a webview for lightbox on Apps) I don't think we'd be able to share the code via this repo.

**Comment by JamieB-gu** [comment]
> The `Image` type, is that supposed to eventually come from the CAPI typedefs?

Good question. Currently DCR and AR transform the structure of the CAPI response to make it easier to work with for their own view code.

To the best of my knowledge DCR's transformations occur in a few specific instances (e.g. [for Photo Essays](https://github.com/guardian/dotcom-rendering/blob/0da0d6063620f9a7024e3f5efba5011a4cab2903/src/model/enhance-photoessay.ts)), and tend to modify bits of the CAPI response in place while sticking to the overall structure. AR, by contrast, parses what information in needs from the CAPI response into an entirely new data structure that in then uses in the rest of the codebase. I don't know exactly what Editions does but you're probably much more familiar with that @AWare.

The AR and DCR teams have a plan to push some of their enhancements they make at the platform-level upstream to CAPI at some point, but I don't have a timeline for that. AR haven't started to look into it because we're still quite focused on building the platform; perhaps @gtrufitt can speak to that for DCR?

In summary, maybe 🙂. If the three main rendering platforms (DCR, AR, Editions) get to the point where they're using the same data type to describe some aspect of a piece of content, in this case `Image`, then I think it's a good candidate for pushing upstream.

**Comment by paperboyo** [comment]
> Do you have any examples of this outside the Lightbox?

Only immersive (hero) images. Eg on top of every gallery. This is where our usual 2 step, 2 dpr, 2 quality values attitude can be least useful (because the initial image size is not CSS-driven, but device/windows size-driven). As discussed with @mxdvl …

**Comment by paperboyo** [comment]
> Lightbox … unless we decide to change that and (e.g. switch to a webview for lightbox on Apps) I don't think we'd be able to share the code via this repo

DCR will have to redo Lightbox one day. Daily just bulit their own. I thought Apps are thinking about it too (or – just done it?), coz Lightboxes were the oldest bits of code. Would be good to have consistent UX and features between products, if not exactly sharing the code upfront.

**Comment by mxdvl** [comment]
In my opinion, it would be perfectly reasonable to expect the return `Img` component to "just work" everywhere. This would only require that a request pass information regarding the expected size of the image in the context it needs to be used. If not passed, a default value of`100vw is used. For example:

- In Lightbox the sizes request would be `(min-width: 61.25rem) calc(100vw-3.75rem), 100vw`
- In an article it would be `(min-width: 61.25rem) 620px, 100vw-3.75rem` (slightly simplified)
- In a full-width hero image it would be `100vw` (which is the default)

**Comment by JamieB-gu** [comment]
> because the initial image size is not CSS-driven, but device/windows size-driven

Isn't this often the case? For example, on mobile body images are typically pretty close to viewport width, so the device size is the determining factor?

**Comment by JamieB-gu** [comment]
> In my opinion, it would be perfectly reasonable to expect the return `Img` component to "just work" everywhere. This would only require that a request pass information regarding the expected size of the image in the context it needs to be used

I like this idea. Taking into account the conversation we've been having about `sizes` offline, perhaps now is the time to define that `Size` type to handle both the HTML `sizes` attribute _and_ the CSS `width` and `height`. An approximation of the code:

```ts
import { css, SerializedStyles } from '@emotion/core';
import { from } from '@guardian/src-foundations/mq';

const enum Breakpoint {
    Mobile,
    Phablet,
    Desktop,
}

type Size = {
    breakpoint: Breakpoint;
    size: string;
}

type Sizes = {
    mediaQueries: Size[];
    default: string;
}

const sizes: Sizes = {
    mediaQueries: [
        { breakpoint: Breakpoint.Phablet, size: '620px' },
    ],
    default: '100vw',
};

const dimensions = (size: string, ratio: number): SerializedStyles => css`
    width: ${size};
    height: calc(${size} * ${ratio});
`;

const styles = (sizes: Sizes, ratio: number): SerializedStyles => css`
    ${dimensions(sizes.default, ratio)}

    ${sizes.mediaQueries.map(query => css`
        ${from[query.breakpoint]} {
            ${dimensions(query.size, ratio)}
        }
    `)}
`;
```

**Comment by paperboyo** [comment]
> on mobile body images are typically pretty close to viewport width, so the device size is the determining factor

Very, very, very good point!

File: package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "image-rendering",
+  "version": "0.1.0",
+  "description": "Handles parsing images from CAPI and rendering them in the *-rendering projects",
+  "main": "index.ts",
+  "dependencies": {
+    "@emotion/core": "^10.0.28",
+    "@guardian/src-foundations": "^2.0.0",
+    "@guardian/types": "github:guardian/types",
```
Comment (webb04) [inline package.json:9]: semver versioning like you added to apps-rendering?

Comment (JamieB-gu) [reply package.json:9]: Good idea!
